src/toxcore/net_crypto/crypto_connection.rs

```diff
@@ -0,0 +1,393 @@
+/*
+    Copyright (C) 2013 Tox project All Rights Reserved.
+    Copyright © 2018 Evgeny Kurnevsky <kurnevsky@gmail.com>
+
+    This file is part of Tox.
+
+    Tox is libre software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Tox is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Tox.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+//! Crypto connection implementation.
+
+use std::net::SocketAddr;
+use std::time::{Duration, SystemTime};
+
+use toxcore::crypto_core::*;
+use toxcore::dht::packet::*;
+
+/// How often in seconds `CookieRequest` or `CryptoHandshake` packets should be
+/// sent
+pub const CRYPTO_SEND_PACKET_INTERVAL: u64 = 1;
+
+/// The maximum number of times we try to send the cookie request and handshake
+/// before giving up
+pub const MAX_NUM_SENDPACKET_TRIES: u8 = 8;
+
+/// If we don't receive UDP packets for this amount of time in seconds the
+/// direct UDP connection is considered dead
+pub const UDP_DIRECT_TIMEOUT: u64 = 8;
+
+/// Packet that should be sent to the peer every second together with info how
+/// many times it was sent and when it was sent last time
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct StatusPacket {
+    /// Packet that should be sent every second. Depending on `ConnectionStatus`
+    /// it can be `CookieRequest` or `CryptoHandshake`
+    pub packet: DhtPacket,
```

Comment:
We'd better have a enum of `CookieRequest` with `CryptoHandshake` instead of DhtPacket